src/modules/rust.rs

```diff
@@ -19,27 +23,85 @@ pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
         return None;
     }
 
-    match get_rust_version() {
-        Some(rust_version) => {
-            const RUST_CHAR: &str = "🦀 ";
+    let module_version = if let Some(toolchain) = env_rustup_toolchain()
+        .or_else(|| execute_rustup_override_list(&context.current_dir))
+        .or_else(|| find_rust_toolchain_file(&context.current_dir))
+    {
+        match execute_rustup_run_rustc_version(&toolchain) {
+            Err(err) => err,
+            Ok(stdout) => format_rustc_version(stdout),
+        }
+    } else {
+        format_rustc_version(execute_rustc_version()?)
+    };
+
+    let mut module = context.new_module("rust");
 
-            let mut module = context.new_module("rust");
-            let module_style = module
-                .config_value_style("style")
-                .unwrap_or_else(|| Color::Red.bold());
-            module.set_style(module_style);
+    let module_style = module
+        .config_value_style("style")
+        .unwrap_or_else(|| Color::Red.bold());
+    module.set_style(module_style);
+    module.new_segment("symbol", RUST_CHAR);
+    module.new_segment("version", &module_version);
 
-            let formatted_version = format_rustc_version(rust_version);
-            module.new_segment("symbol", RUST_CHAR);
-            module.new_segment("version", &formatted_version);
+    Some(module)
+}
 
-            Some(module)
+fn env_rustup_toolchain() -> Option<String> {
+    let val = env::var("RUSTUP_TOOLCHAIN").ok()?;
+    Some(val.trim().to_owned())
+}
+
+fn execute_rustup_override_list(cwd: &Path) -> Option<String> {
+    let Output { stdout, .. } = Command::new("rustup")
+        .args(&["override", "list"])
+        .output()
+        .ok()?;
+    let stdout = String::from_utf8(stdout).ok()?;
+    if stdout == "no overrides\n" {
+        return None;
+    }
+    stdout
+        .lines()
+        .flat_map(|line| {
+            let mut words = line.split_whitespace();
+            let dir = words.next()?;
+            let toolchain = words.next()?;
+            Some((dir, toolchain))
+        })
+        .find(|(dir, _)| cwd.starts_with(dir))
+        .map(|(_, toolchain)| toolchain.to_owned())
+}
+
+fn find_rust_toolchain_file(mut dir: &Path) -> Option<String> {
+    loop {
```

Comment:
Should we limit the maximum depth here to some small reasonable number? Though I realise this will potentially compromise accuracy..

Reply:
I don't think this degrades the performance. Its behavior is the same as [`~/.cargo/bin/rustc --version`](https://github.com/rust-lang/rustup.rs/blob/d84e6e50126bccd84649e42482fc35a11d019401/src/config.rs#L320-L358
).

Reply:
My benchmarks suggests this is about 30% slower. That said, I think it is quick enough - under 100ms on most systems and under 50ms on systems with fast I/O.